### PR TITLE
Add "Group by" mode

### PR DESCRIPTION
**Pull request by dsmilkov**
https://github.com/lilacai/lilac/assets/2294279/a57659e2-ba83-4790-804c-e0155e2d2389

### CODE REVIEW

File: lilac/data/dataset.py [hunk 1]
```diff
@@ -297,7 +295,17 @@ class ConceptSearch(BaseModel):
...
 
-Search = Union[ConceptSearch, SemanticSearch, KeywordSearch]
+class MetadataSearch(BaseModel):
+  """A metadata search query on a column."""
+  path: Path
+  op: FilterOp
+  value: Optional[Union[FeatureValue, FeatureListValue]] = None
+  type: Literal['metadata'] = 'metadata'
+
+  model_config = ConfigDict(json_schema_extra=_fix_const_in_schema('type', 'metadata'))
```
Comment (nsthorat) [inline lilac/data/dataset.py:305]: can we rename this "_fix_const_in_schema" to what it does specifically

File: lilac/signals/metadata_op_search.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+"""A signal to filter metadata based on an operation."""
```
Comment (nsthorat) [inline lilac/signals/metadata_op_search.py:1]: Can you add a big note here saying this is very internal

**Review by nsthorat — APPROVED** [review]

File: lilac/data/dataset.py (hunk 1 above)
Comment (dsmilkov) [reply lilac/data/dataset.py:305]: done

File: lilac/signals/metadata_op_search.py (hunk 2 above)
Comment (dsmilkov) [reply lilac/signals/metadata_op_search.py:1]: done

File: lilac/signals/metadata_op_search.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  start = 0
+  while True:
+    start = text.find(subtext, start)
+    if start == -1:
+      return
+    yield start, start + subtext_len
+    start += subtext_len
+
+
+class MetadataOpSignal(Signal):
```
Comment (nsthorat) [inline lilac/signals/metadata_op_search.py:25]: FilterMaskSignal?

Comment (dsmilkov) [reply lilac/signals/metadata_op_search.py:11]: done

**Pull request merged**